src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -264,6 +266,12 @@ protected Object invokeStep(StepDescriptor d, String name, Object args) {
 
             // Persist the node - block start and end nodes do their own persistence.
             CpsFlowExecution.maybeAutoPersistNode(an);
+
+            // Call any registered StepListeners.
+            for (StepListener sl : ExtensionList.lookup(StepListener.class)) {
+                sl.newStep(s, context);
```

Comment:
If a `StepListener` extension is broken for some reason, then just by throwing an exception it could stop all steps from executing. We catch `Throwable` around calls to `GraphListener` and log failures, but if we intend for the listeners to throw exceptions to stop the step, that wouldn't work here. 

Maybe it would be better to have `StepListener` implementations call `context.onFailure` when they want to abort a step, rather than throwing an exception directly out of `StepListener.newStep`, so that we could distinguish "this listener is broken" from "this listener wants to stop the step from executing". You'd need to tweak the code here to look at `CpsStepContext.isCompleted` before calling `s.start(context)` in that case, and I'm not sure whether we'd want to run all listeners or not once one failed the step.

Maybe also worth documenting upstream that you should not call `StepContext.onSuccess` or `StepContext.newBodyInvoker` (or `StepContext.onFailure` with the current code) from a listener, or wrapping the context here to make those methods throw exceptions?

One other minor note is that `Extension.ordinal` seems somewhat relevant for this extension, since any listener that just wants to collect data/metrics about steps (if that is a valid use case) but probably wants consistency in whether they run before or after extensions that might fail steps. If `CpsStepContext.onFailure` were used, then maybe this wouldn't matter if we always ran all listeners, although then we'd have the possibility that more than one listener might try to fail a step which would cause some "already completed" warnings in `CpsStepContext.completed`.

Reply:
Oh, duh, `context.onFailure` is the right way to do it. I'll rework this.